rfcs/20191212-keras-categorical-inputs.md

```diff
@@ -0,0 +1,390 @@
+# Keras categorical inputs
+
+| Status        | Proposed                                             |
+:-------------- |:---------------------------------------------------- |
+| **Author(s)** | Zhenyu Tan (tanzheny@google.com), Francois Chollet (fchollet@google.com)|
+| **Sponsor**   | Karmel Allison (karmel@google.com), Martin Wicke (wicke@google.com) |
+| **Updated**   | 2019-12-12                                           |
+
+## Objective
+
+This document proposes 4 new preprocessing Keras layers (`CategoryLookup`, `CategoryCrossing`, `CategoryEncoding`, `CategoryHashing`), and 1 additional op (`to_sparse`) to allow users to:
+* Perform feature engineering for categorical inputs
+* Replace feature columns and `tf.keras.layers.DenseFeatures` with proposed layers
+* Introduce sparse inputs that work with Keras linear models and other layers that support sparsity
+
+Other proposed layers for replacement of feature columns such as `tf.feature_column.bucketized_column` and `tf.feature_column.numeric_column` has been discussed [here](https://github.com/keras-team/governance/blob/master/rfcs/20190502-preprocessing-layers.md) and are not the focus of this document.
+
+## Motivation
+
+Specifically, by introducing the 4 layers, we aim to address these pain points:
+* Users have to define both feature columns and Keras Inputs for the model, resulting in code duplication and deviation from DRY (Do not repeat yourself) principle. See this [Github issue](https://github.com/tensorflow/tensorflow/issues/27416).
+* Users with large dimension categorical inputs will incur large memory footprint and computation cost, if wrapped with indicator column through `tf.keras.layers.DenseFeatures`.
+* Currently there is no way to correctly feed Keras linear model or dense layer with multivalent categorical inputs or weighted categorical inputs.
+
+## User Benefit
+
+We expect to get rid of the user painpoints once migrating off feature columns.
+
+## Example Workflows
+
+Two example workflows are presented below. These workflows can be found at this [colab](https://colab.sandbox.google.com/drive/1cEJhSYLcc2MKH7itwcDvue4PfvrLN-OR#scrollTo=22sa0D19kxXY).
+
+### Workflow 1
```

Comment:
Would it make sense to provide an example workflow where you have to get the vocabulary from e.g. a csv file using tf.data and in particular, tf.data.experimental.unique and tf.data.experimental.get_single_element to read out the tensor?  @jsimsa wdyt?

This is gonna be very common, i think, in real use cases.